File: katas/javascript/interactive-coding-challenges/test/CompressStringTest.js
const assert = require('assert');
const CompressString = require('../src/CompressString');

describe('CompressString', () => {
  describe('returns compressed string', () => {
    it('\'AAABCCDDDD\' returns \'A3BC2D4\'', () => {
      assert.equal(CompressString('AAABCCDDDD'), 'A3BC2D4');
    });

    it('\'AAABCCDDDDE\' returns \'A3BC2D4E\'', () => {
      assert.equal(CompressString('AAABCCDDDDE'), 'A3BC2D4E');
    });

    it('\'BAAACCDDDD\' returns \'BA3C2D4\'', () => {
      assert.equal(CompressString('BAAACCDDDD'), 'BA3C2D4');
    });

    it('\'AAABAACCDDDD\' returns \'A3BA2C2D4\'', () => {
      assert.equal(CompressString('AAABAACCDDDD'), 'A3BA2C2D4');
    });
  });

  describe('returns original string if compression does not save space', () => {
    it('\'\' returns \'\'', () => {
      assert.equal(CompressString(''), '');
    });

    it('\'aa\' returns \'aa\'', () => {
      assert.equal(CompressString('aa'), 'aa');
    });

    it('\'AABBCC\' returns \'AABBCC\'', () => {
      assert.equal(CompressString('AABBCC'), 'AABBCC');
    });
  });
});

File: katas/java/src/test/java/interactive_coding_challenges/arrays_and_strings/CompressStringTest.java
package interactive_coding_challenges.arrays_and_strings;

import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

/**
 * @author Mike Lowe
 */
public final class CompressStringTest {

  @Test
  public void compress_null_returnsNull() {
    assertNull(CompressString.compress(null));
  }

  @Test
  public void compress_empty_returnsEmpty() {
    assertEquals("", CompressString.compress(""));
  }

  @Test
  public void compress_redundantCompression_returnsOriginalString() {
    assertEquals("a", CompressString.compress("a"));
    assertEquals("aa", CompressString.compress("aa"));
    assertEquals("ab", CompressString.compress("ab"));
    assertEquals("aab", CompressString.compress("aab"));
    assertEquals("abc", CompressString.compress("abc"));
    assertEquals("aabbc", CompressString.compress("aabbc"));
    assertEquals("aabbcc", CompressString.compress("aabbcc"));
  }

  @Test
  public void compress_needsCompressing_returnsCompressed() {
    assertEquals("a3bc2d4e", CompressString.compress("aaabccdddde"));
    assertEquals("ba3c2d4", CompressString.compress("baaaccdddd"));
    assertEquals("a3ba2c2d4", CompressString.compress("aaabaaccdddd"));
  }

}
